clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -305,6 +303,9 @@ private void writeDefaultBatchHeader() {
                 partitionLeaderEpoch, numRecords);
 
         buffer.position(pos);
+
+        // update the compression ratio
+        this.compressionRatio = (float) writtenCompressed / this.writtenUncompressed;
```

Comment:
I find it slightly weird to update the ratio here. Yes it is the last operation, but clearer to do it outside (in the `close` method itself) no?